x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/CommunityIdProcessor.java

```diff
@@ -0,0 +1,571 @@
+            return result < 0 || (result == 0 && sourcePort < destinationPort);
+        }
+
+        byte[] toBytes() {
+            boolean hasPort = TRANSPORTS_WITH_PORTS.contains(protocol);
+            int len = source.getAddress().length + destination.getAddress().length + 2 + (hasPort ? 4 : 0);
+            ByteBuffer bb = ByteBuffer.allocate(len);
+
+            boolean isOneWay = false;
+            if (protocol == Transport.Icmp || protocol == Transport.IcmpIpV6) {
+                // ICMP protocols populate port fields with ICMP data
+                Integer equivalent = IcmpType.codeEquivalent(icmpType, protocol == Transport.IcmpIpV6);
+                isOneWay = equivalent == null;
+                sourcePort = icmpType;
+                destinationPort = equivalent == null ? icmpCode : equivalent;
+            }
+
+            boolean keepOrder = isOrdered() || ((protocol == Transport.Icmp || protocol == Transport.IcmpIpV6) && isOneWay);
+            bb.put(keepOrder ? source.getAddress() : destination.getAddress());
+            bb.put(keepOrder ? destination.getAddress() : source.getAddress());
+            bb.put(toUint16(protocol.getTransportNumber() << 8));
+
+            if (hasPort) {
+                bb.put(keepOrder ? toUint16(sourcePort) : toUint16(destinationPort));
+                bb.put(keepOrder ? toUint16(destinationPort) : toUint16(sourcePort));
+            }
+
+            return bb.array();
+        }
+
+        String toCommunityId(MessageDigest md, byte[] seed) {
+            md.update(seed);
+            byte[] encodedBytes = Base64.getEncoder().encode(md.digest(toBytes()));
+            return "1:" + new String(encodedBytes, StandardCharsets.UTF_8);
+        }
+    }
+
+    public enum Transport {
+        Icmp(1),
+        Igmp(2),
+        Tcp(6),
+        Udp(17),
+        Gre(47),
+        IcmpIpV6(58),
+        Eigrp(88),
+        Ospf(89),
+        Pim(103),
+        Sctp(132);
+
+        private final int transportNumber;
+
+        private static final Map<String, Transport> TRANSPORT_NAMES;
+
+        static {
+            TRANSPORT_NAMES = new HashMap<>();
+            TRANSPORT_NAMES.put("icmp", Icmp);
+            TRANSPORT_NAMES.put("igmp", Igmp);
+            TRANSPORT_NAMES.put("tcp", Tcp);
+            TRANSPORT_NAMES.put("udp", Udp);
+            TRANSPORT_NAMES.put("gre", Gre);
+            TRANSPORT_NAMES.put("ipv6-icmp", IcmpIpV6);
+            TRANSPORT_NAMES.put("icmpv6", IcmpIpV6);
+            TRANSPORT_NAMES.put("eigrp", Eigrp);
+            TRANSPORT_NAMES.put("ospf", Ospf);
+            TRANSPORT_NAMES.put("pim", Pim);
+            TRANSPORT_NAMES.put("sctp", Sctp);
+        }
+
+        Transport(int transportNumber) {
+            this.transportNumber = transportNumber;
+        }
+
+        public int getTransportNumber() {
+            return transportNumber;
+        }
+
+        public static Transport fromNumber(int transportNumber) {
```

Comment:
Not sure if this is a worth it, so feel free to ignore. To ensure that `fromNumber` stays up-to-date with all the enum transportNumbers it might be useful to have test that checks the all the `Transport.values()` work with `fromNumber` without exception.